src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxListPool.cs

```diff
@@ -2,24 +2,43 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+// #define LOG
+
 using System;
+#if LOG
 using System.Collections.Generic;
 using System.Diagnostics;
-using Roslyn.Utilities;
+#endif
+using Microsoft.CodeAnalysis.PooledObjects;
 
 namespace Microsoft.CodeAnalysis.Syntax.InternalSyntax
 {
     internal class SyntaxListPool
     {
-        private ArrayElement<SyntaxListBuilder?>[] _freeList = new ArrayElement<SyntaxListBuilder?>[10];
+        private static readonly ObjectPool<SyntaxListPool> s_listPool = new ObjectPool<SyntaxListPool>(() => new SyntaxListPool());
+
+        private const int InitialFreeListSize = 16;
+        private const int InitialBuilderCapacity = 32;
+
+        private ArrayElement<SyntaxListBuilder?>[] _freeList = new ArrayElement<SyntaxListBuilder?>[InitialFreeListSize];
         private int _freeIndex;
 
-#if DEBUG
+#if LOG
```

Comment:
This is useful, but only when you are really looking into this code. The overhead seems high enough that it's not worth doing unless it's being investigated, thus the removal from vanilla debug builds.